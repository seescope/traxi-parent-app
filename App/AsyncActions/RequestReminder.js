// @flow
import Analytics from 'react-native-analytics';
import OneSignal from 'react-native-onesignal';

import type { SetupState } from '../Reducers/Setup';
import type { KidsState } from '../Reducers/Kids';
import type { ParentState } from '../Reducers/Parent';

type Dispatch = () => void;
type RootState = {
  kidsState: KidsState,
  parentState: ParentState,
  setupState: SetupState
};
type GetState = () => RootState;

export default () =>
  (dispatch: Dispatch, getState: GetState) => {
    const { parentState, setupState, kidsState } = getState();
    const { email, UUID } = parentState;
    const { kidUUID } = setupState;

    if (!email) return Promise.reject('No email specified!');
    if (!kidUUID) return Promise.reject('No kid UUID!');

    const { name } = kidsState[kidUUID];

    Analytics.track('Reminder requested', {
      kidName: name,
    });
    Analytics.identify(UUID, { kidName: name, email });
    OneSignal.sendTags({
      UUID,
      kidName: name,
      email,
    });

    return Promise.resolve({
      type: 'REMINDER_REQUESTED',
    });
  };
