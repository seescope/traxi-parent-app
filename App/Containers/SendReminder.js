import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Text, View, Alert } from 'react-native';
import { Actions } from 'react-native-router-flux';

import requestReminder from '../AsyncActions/RequestReminder';
import Button from '../Components/Button';
import Background from '../Components/Background';
import HeaderText from '../Components/HeaderText';
import Spacing from '../Components/Spacing';
import { GREY } from '../Constants/Colours';
import { firstName } from '../Utils';
import STYLES from '../Constants/Styles';

const style = {
  container: {
    marginTop: 32,
    paddingTop: 18,
    alignItems: 'center',
  },
  cardContainer: {
    marginTop: 32,
    paddingTop: 18,
  },
  labelText: {
    textAlign: 'left',
    fontFamily: 'Raleway-Regular',
    fontSize: 14,
    color: GREY,
  },
  innerContainer: {
    paddingHorizontal: 16,
  },
  buttonsContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
};

const SendReminder = ({ kidName, onPress }) => (
  <Background>
    <View style={style.container}>
      <HeaderText>Get a reminder</HeaderText>

      <Spacing height={16} />

      <View style={[STYLES.CARD, style.container]} elevation={6}>
        <Text style={style.labelText}>
          Not ready yet? No problem!
        </Text>

        <Spacing height={16} />

        <Text style={style.labelText}>
          Tap the button below and Traxi will send you a reminder to set up
          {' '}
          {kidName}
          â€™s device tomorrow.
        </Text>

        <Spacing height={16} />
      </View>

      <Spacing height={32} />

      <View style={style.buttonsContainer}>
        <Button onPress={() => Actions.pop()}>
          Back
        </Button>
        <Button primary onPress={onPress}>
          Remind me tomorrow
        </Button>
      </View>
    </View>
  </Background>
);
SendReminder.propTypes = {
  kidName: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired,
};
const mapStateToProps = state => {
  const { kidUUID } = state.setupState;
  const { email } = state.parentState;
  const { name } = state.kidsState[kidUUID] || {};
  return {
    email,
    kidUUID,
    kidName: firstName(name) || '',
  };
};
export const mergeProps = ({ email, kidName }, { dispatch }) => ({
  kidName,
  onPress: async () => {
    await dispatch(requestReminder());
    return Alert.alert(
      'Reminder requested',
      `A reminder will be sent to ${email} tomorrow.`,
      [
        {
          text: 'OK',
          onPress: () => Actions.pop(),
        },
      ]
    );
  },
});
export default connect(mapStateToProps, null, mergeProps)(SendReminder);
