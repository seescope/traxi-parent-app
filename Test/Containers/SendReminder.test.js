import React from 'react';
import renderer from 'react-test-renderer';
import { Alert } from 'react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
jest.mock('../../App/AsyncActions/RequestReminder', () =>
  jest.fn(email => Promise.resolve({ type: 'REMINDER_REQUESTED', email })));
import RequestReminder from '../../App/AsyncActions/RequestReminder';

import SendReminder, { mergeProps } from '../../App/Containers/SendReminder';

const mockStore = configureStore([thunk]);
const testStore = mockStore({
  parentState: {
    email: 'some@email.com',
  },
  kidsState: {
    'abc-123': {
      name: 'John Bobson',
    },
  },
  setupState: {
    kidUUID: 'abc-123',
  },
});

const SendReminderComponent = () => (
  <Provider store={testStore}>
    <SendReminder />
  </Provider>
);

it('renders the <SendReminder> component', () => {
  const tree = renderer.create(<SendReminderComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('sends a reminder then shows an alert', () => {
  Alert.alert = jest.fn();
  const mockDispatch = jest.fn(v => v);

  const { onPress } = mergeProps(
    { email: 'test@email.com' },
    { dispatch: mockDispatch }
  );

  return onPress().then(() => {
    expect(RequestReminder).toHaveBeenCalled();
    expect(Alert.alert).toHaveBeenCalled();
  });
});
