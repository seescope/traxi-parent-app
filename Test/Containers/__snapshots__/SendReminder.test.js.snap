// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the <SendReminder> component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F2F7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginTop": 32,
        "paddingTop": 18,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "#545454",
          "fontFamily": "Raleway-ExtraBold",
          "fontSize": 26,
          "textAlign": "center",
        }
      }
    >
      Get a reminder
    </Text>
    <View
      style={
        Object {
          "height": 16,
        }
      }
    />
    <View
      elevation={6}
      style={
        Array [
          Object {
            "backgroundColor": "#FFF",
            "borderRadius": 4,
            "elevation": 6,
            "margin": 8,
            "marginBottom": 24,
            "padding": 8,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 3,
            },
            "shadowOpacity": 0.16,
            "shadowRadius": 3,
            "width": 734,
          },
          Object {
            "alignItems": "center",
            "marginTop": 32,
            "paddingTop": 18,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#545454",
            "fontFamily": "Raleway-Regular",
            "fontSize": 14,
          }
        }
      >
        Not ready yet? Enter your email address below and we’ll send you reminder for you to finish setting up
         
        John
        ’s device tomorrow.
      </Text>
      <View
        style={
          Object {
            "height": 16,
          }
        }
      />
      <View>
        <TextInput
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="email-address"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholderTextColor="rgba(75, 73, 73, 0.8)"
          secureTextEntry={undefined}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderRadius": 4,
              "color": "rgba(75, 73, 73, 1)",
              "fontFamily": "Raleway-Regular",
              "height": 32,
              "marginTop": 4,
              "paddingHorizontal": 8,
              "width": 686,
            }
          }
          underlineColorAndroid="#448AFF"
          value={undefined}
        />
        <View
          style={
            Object {
              "backgroundColor": "#448AFF",
              "height": 2,
              "marginBottom": 16,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "height": 32,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <View
        style={
          Object {
            "marginHorizontal": 16,
          }
        }
      >
        <View
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#448AFF",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#FFF",
                "fontFamily": "Raleway-Bold",
              }
            }
          >
            Remind me tomorrow
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
